### PR TITLE
Importer state

### PR DESCRIPTION
**Pull request by simplicbe**
Add a state option for import modules, to get rid of throwing exceptions if it the module not be used for importing some other modules or packages.

**EDIT**
Note, everything should be backward compatible, so it won't cause any problems.

### CODE REVIEW

**Comment by slide** [comment]
I'm not sure this works in all cases. Some code is expecting an exception.

**Comment by simplicbe** [comment]
Do you have any example? So we may could find a good, general solution. Thank you for your feedback.

**Comment by slide** [comment]
I would need to go back through some of the testing I had done previously, this was back a long time ago when I implemented zipimport.

File: Languages/IronPython/IronPython/Runtime/Importer.cs [hunk 1]
```diff
@@ -917,8 +917,19 @@ class PythonFileStreamContentProvider : StreamContentProvider {
                 try {
                     object handler = PythonCalls.Call(context, hook, dirname);
 
-                    if (handler != null) {
-                        return handler;
+                    if (handler != null)
+                    {
+                        if (handler is IImporterModule)
+                        {
+                            if ((handler as IImporterModule).State == ImporterModuleState.Ready)
```
Comment (jdhardy) [inline Languages/IronPython/IronPython/Runtime/Importer.cs]: Use an `((IImporterModule)handler).State; you've already checked with `is`, no need to check again with `as`.

**Comment by jdhardy** [comment]
Can you give an example of what problem this is trying to solve? And what does CPython do in this case?

**Comment by simplicbe** [comment]
I'm not sure how to proof, how python act in this case. I just noticed than when i debug an application, the `Importer.cs` often throws exceptions, which is in my point of view not the best soltuion, because throwing and catching exceptions is slower than working with some state based system.

Hope i'm not totally wrong.

**Comment by slide** [comment]
Python raises an ImportError like the code currently does. I know because I ported the code from Python for zipimport.

**Comment by rassilon** [comment]
I think the only sane way for not throwing exceptions in these cases would be to add a INoExceptionImporter (or a more pretty bikeshed color of your choosing) interface that the path hook loop could use instead.

That way, the loop would hopefully be able to conceptually check: (hook as PythonType).UnderlyingSystemType.ImplementedInterfaces contains INoExceptionImporter and thus skip the PythonCalls.Call invocation.

I'm just not certain atm, what the correct approach (i.e. the actual method calls/checks on PythonType) would be (if there is one) that would yield the necessary INoExceptionImporter interface from zipimporter. (Aside from special casing ONLY the C# zipimporter in this path hook loop.)

I agree with the other folks, that this particular approach just isn't compatible with direct code usages of zipimporter (outside of IronPython's import hook system).

i.e. What if someone wrote an import hook that internally called zipimporter? This PR would break that importer.

**Comment by DinoV** [comment]
Another sane way would be to use the existing LightException support.  Instead of throwing an actual exception the importer can return LightException objects and existing callers would need to check for them.  Any calls which leak out into Python code would need to be decorated with the LightThrows (?)attribute.  Then we wouldn't actually be throwing any .NET exceptions including into Python code which catches ImportErrors, but if an exception ever hit a boundary where it needed to be thrown (for example across some .NET code which doesn't understand light exceptions) then it would get transformed into a .NET exception.

**Comment by simplicbe** [comment]
How to return a LightException if it will be "raised" in the constructor?

**Comment by DinoV** [comment]
The only way to do that would be to modify the type to have __new__ / __init__ methods and return it from one of those.  An example of a type which does that (but not for light exceptions) is PythonType or OldClass.  If it needs to be constructed from C# and still potentially throw the C# code would have to be updated to call __new__/__init__ and check for errors.  When called from Python we'll generate the correct sequence of __new__/__init__ as if they were implemented on a type in Python it's self.

**Comment by slide** [comment]
@DinoV Do you have to provide both __new__ and __init__ and can either method return a value, or only __new__?

**Comment by DinoV** [comment]
Only __new__ can return a new value, __init__ should always be typed to void / return None.  I don't think you need to provide both because I think we should pick up object.__init__ if it's not defined, and we should allow *args and **kwargs to be passed to it.  But the rules are a little subtle there so I could be getting that last bit wrong.

**Comment by slide** [comment]
I'm going to close this PR and request that if someone wants to do this, they implement the LightException way of doing it.

**Pull request closed without merging**